pkg/sample/random_test.go

```diff
@@ -0,0 +1,867 @@
+		},
+		// Case 026
+		{
+			a: 5,
+			b: 7,
+		},
+
+		// Case 027
+		{
+			a: 6,
+			b: 1,
+		},
+		// Case 028
+		{
+			a: 1,
+			b: 6,
+		},
+
+		// Case 029
+		{
+			a: 30,
+			b: 5,
+		},
+		// Case 030
+		{
+			a: 5,
+			b: 30,
+		},
+
+		// Case 031
+		{
+			a: 57,
+			b: 2,
+		},
+		// Case 032
+		{
+			a: 2,
+			b: 57,
+		},
+
+		// Case 033
+		{
+			a: 108,
+			b: 20,
+		},
+		// Case 034
+		{
+			a: 20,
+			b: 108,
+		},
+	}
+
+	for i, tc := range testCases {
+		t.Run(fmt.Sprintf("%03d", i), func(t *testing.T) {
+			mpx := map[string]int{}
+			mpy := map[string]int{}
+
+			for i := 0; i < 1000; i++ {
+				var sam *Sample
+				{
+					sam = New(Config{
+						Rea: rand.Reader,
+					})
+				}
+
+				var x []int
+				var y []int
+				{
+					x, y = sam.Random(tc.a, tc.b)
+				}
+
+				{
+					mpx[musStr(x)]++
+					mpy[musStr(y)]++
+				}
+
+				// If we have a positive line on both sides, then the number of random
+				// points given to us in return should always be equal.
+				if tc.a != 0 && tc.b != 0 && len(x) != len(y) {
+					t.Fatalf("expected %#v got %#v", len(x), len(y))
+				}
+
+				{
+					if generic.Duplicate(x) {
+						t.Fatalf("expected x to not contain duplicates")
+					}
+
+					if generic.Duplicate(y) {
+						t.Fatalf("expected y to not contain duplicates")
+					}
+				}
+
+				// Below we allocate slices to access items using the generated indices.
+				// If we are able to access those items using any given index without
+				// creating a runtime panic, then our implementation is safe to the
+				// extend that no indices are generated out of bounds.
+				{
+					slx := make([]int, tc.a)
+					for _, i := range x {
+						_ = slx[i]
+					}
+
+					sly := make([]int, tc.b)
+					for _, i := range y {
+						_ = sly[i]
+					}
+				}
+			}
+
+			// We somehow want to assert that we generate a variation of random
+			// indices. Below we check that there are some minimum amounts of
+			// generated permutations, which implies that each of the existing keys in
+			// the associated map of observed indices was generated at least once.
+			// More keys equals more randomness. And so we compare against the
+			// smallest common denominator of the given test case.
+			{
+				lnx := uint(len(mpx))
+				mnx := min(tc.a, tc.b)
+				if lnx < mnx {
+					t.Fatalf("expected x to generate at least %d different variations, got %d", lnx, mnx)
+				}
+
+				lny := uint(len(mpy))
+				mny := min(tc.a, tc.b)
+				if lny < mny {
+					t.Fatalf("expected y to generate at least %d different variations, got %d", lny, mny)
+				}
+			}
+		})
+	}
+}
+
+func Test_Sample_Random_Fake(t *testing.T) {
+	testCases := []struct {
+		a uint
+		b uint
+		s int
+		x []int
+		y []int
+	}{
+		// Case 000
+		{
+			a: 0,
+			b: 0,
+			s: 0,
+			x: nil,
+			y: nil,
+		},
+		// Case 001
+		{
+			a: 0,
+			b: 0,
+			s: 3,
+			x: nil,
+			y: nil,
+		},
+
+		// Case 002
+		{
+			a: 1,
+			b: 0,
+			s: 0,
+			x: []int{
+				0,
+			},
+			y: nil,
+		},
```

Comment:
This test here shows that there is only a single address to select if there is only 1 staker.